File: linkedlist/reverse_linkedlist.py
class Node: 
    def __init__(self, data):
        self.data = data 
        self.next = None
         

def display(head: Node) -> None:
    temp = head 
    while temp: 
        print(f"{temp.data} -> ", end=" ")
        temp = temp.next 
    print()

def populate() -> Node:
    head = Node(1)
    temp = head 
    for i in range(2, 10):
        temp.next = Node(i) 
        temp = temp.next 
    return head

def reverse_recursive(head: Node) -> Node: 
    #base case
    if head is None or head.next is None: 
        return head 
    small_head = reverse_recursive(head.next) 
    head.next.next, head.next = head, None 
    return small_head


def reverse(head: Node) -> Node: 
    #empty check 
    if head is None: 
        return None
     
    prev, curr, forward = None, head, None 
    while(curr):
        forward = curr.next 
        curr.next = prev 
        prev = curr 
        curr = forward 
    return prev #as curr and forward will be None at last 

if __name__ == '__main__':
    link = populate()
    display(head=link)
    # reverse_link = reverse(link)
    reverse_link = reverse_recursive(link)
    display(head=reverse_link)

File: linkedlist/reverse_linkedlist.cpp
#include<iostream>
using namespace std; 

class Node{
    public:
        int data; 
        Node* next; 

        Node(int data){
            this -> data = data; 
            this -> next = NULL; 
        }
};

Node* reverse(Node* head){
    Node* prev = NULL;  
    Node* forward = NULL;
    Node* curr = head;  
    while(curr){
        forward = curr -> next;  
        curr ->  next = prev; 
        prev = curr; 
        curr = forward; 
    }
    return prev; 

}
void print(Node* head){
    Node* temp = head; 
    while (temp){
        cout << temp -> data << "->";
        temp = temp-> next;  
    }
    cout  << endl;
}

Node* populate(){
    Node* head = new Node(-1);
    Node* Main = head; 
    for(int i = 0; i < 10; i++){
        Node* temp = new Node(i); 
        head -> next = temp; 
        head = temp; 
    }
    return Main -> next; 
}
int main(){
    Node* head = populate();
    print(head);
    Node* reverse_head = reverse(head);
    print(reverse_head);
    return 0;
}